src/pyclass.rs

```diff
@@ -333,6 +348,46 @@ fn py_class_method_defs<T: PyMethods>() -> (
     (new, call, defs)
 }
 
+/// Generates the __dictoffset__ and __weaklistoffset__ members, to set tp_dictoffset and
+/// tp_weaklistoffset.
+///
+/// Only works on Python 3.9 and up.
+#[cfg(Py_3_9)]
+fn py_class_members<T: PyClass>() -> Vec<ffi::structmember::PyMemberDef> {
+    static DICTOFFSET: &str = "__dictoffset__\0";
+    static WEAKLISTOFFSET: &str = "__weaklistoffset__\0";
+
+    let mut members = Vec::new();
+
+    // __dict__ support
+    if let Some(dict_offset) = PyCell::<T>::dict_offset() {
+        members.push(ffi::structmember::PyMemberDef {
+            name: DICTOFFSET.as_ptr() as _,
+            type_code: ffi::structmember::T_PYSSIZET,
+            offset: dict_offset as _,
+            flags: ffi::structmember::READONLY,
+            doc: std::ptr::null_mut(),
+        });
+    }
+
+    // weakref support
+    if let Some(weakref_offset) = PyCell::<T>::weakref_offset() {
+        members.push(ffi::structmember::PyMemberDef {
+            name: WEAKLISTOFFSET.as_ptr() as _,
+            type_code: ffi::structmember::T_PYSSIZET,
+            offset: weakref_offset as _,
+            flags: ffi::structmember::READONLY,
+            doc: std::ptr::null_mut(),
+        });
+    }
```

Comment:
These two `PyMemberDef` have many common fields and you might want to shorten the code by using a closure or function.

Reply:
I'll use a macro (so that the optimizer doesn't have to try to eliminate the function call overhead).

Reply:
How about `const fn`?

Reply:
`const fn` won't work here, because the `T: PyClass` generic type bound is unstable with `const fn`.

Reply:
I just meant:
```rust
pub const fn member_def(name: *mut c_char, offset: Py_ssize_t) -> PyMemberDef {
    PyMemberDef {
        name,
        type_code: crate::ffi::structmember::T_PYSSIZET,
        offset,
        flags: crate::ffi::structmember::READONLY,
        doc: std::ptr::null_mut(),
    }
}
```